components/script/dom/gamepadhapticactuator.rs

```diff
@@ -0,0 +1,381 @@
+                    ));
+                    return playing_effect_promise;
+                } else if *params.weakMagnitude < 0.0 || *params.weakMagnitude > 1.0 {
+                    playing_effect_promise.reject_error(Error::Type(
+                        "Weak magnitude value is not within range of 0.0 to 1.0.".to_string(),
+                    ));
+                    return playing_effect_promise;
+                } else if *params.leftTrigger < 0.0 || *params.leftTrigger > 1.0 {
+                    playing_effect_promise.reject_error(Error::Type(
+                        "Left trigger value is not within range of 0.0 to 1.0.".to_string(),
+                    ));
+                    return playing_effect_promise;
+                } else if *params.rightTrigger < 0.0 || *params.rightTrigger > 1.0 {
+                    playing_effect_promise.reject_error(Error::Type(
+                        "Right trigger value is not within range of 0.0 to 1.0.".to_string(),
+                    ));
+                    return playing_effect_promise;
+                }
+            },
+        }
+
+        let document = self.global().as_window().Document();
+        if !document.is_fully_active() {
+            playing_effect_promise.reject_error(Error::InvalidState);
+        }
+
+        self.sequence_id.set(self.sequence_id.get().wrapping_add(1));
+
+        if self.playing_effect_promise.borrow().is_some() {
+            let trusted_promise = TrustedPromise::new(
+                self.playing_effect_promise
+                    .borrow()
+                    .clone()
+                    .expect("Promise is null!"),
+            );
+            *self.playing_effect_promise.borrow_mut() = None;
+            let _ = self.global().gamepad_task_source().queue(
+                task!(preempt_promise: move || {
+                    let promise = trusted_promise.root();
+                    let message = DOMString::from("preempted");
+                    promise.resolve_native(&message);
+                }),
+                &self.global(),
+            );
+        }
+
+        if !self.effects.contains(&type_) {
+            playing_effect_promise.reject_error(Error::NotSupported);
+        }
+
+        assert!(self.playing_effect_promise.borrow().is_none());
+
+        *self.playing_effect_promise.borrow_mut() = Some(playing_effect_promise.clone());
+        self.effect_sequence_id.set(self.sequence_id.get());
+
+        // XXX The spec says we SHOULD also pass a playEffectTimestamp for more precise playback timing
+        // when start_delay is non-zero, but this is left more as a footnote without much elaboration.
+
+        let params = DualRumbleEffectParams {
+            duration: params.duration as f64,
+            start_delay: params.startDelay as f64,
+            strong_magnitude: *params.strongMagnitude,
+            weak_magnitude: *params.weakMagnitude,
+        };
+        let event = EmbedderMsg::PlayGamepadHapticEffect(
```

Comment:
For consistency and ease of reading, I think it would be good to use the IPC router to handle the reponse, as opposed to go through the globalscope. As noted elsewhere, you don't need the extra message to the constellation.